[Datasets] ObjectFetchTimedOutError with dynamic block splitting enabled

### What happened + What you expected to happen

With https://github.com/ray-project/ray/pull/28841, CI test:

* [python/ray/data/tests/test_dataset_text.py::test_read_text_remote_args](https://github.com/ray-project/ray/blob/master/python/ray/data/tests/test_dataset_text.py#L113)

Failed with error

```
python/ray/data/tests/test_dataset_text.py::test_read_text_remote_args FAILED                                                                            [100%]

=========================================================================== FAILURES ===========================================================================
__________________________________________________________________ test_read_text_remote_args __________________________________________________________________

ray_start_cluster = <ray.cluster_utils.Cluster object at 0x7fbc3b152850>
tmp_path = PosixPath('/private/var/folders/m6/0xsy6zks1z9c9dc0v1pthsyr0000gn/T/pytest-of-chengsu/pytest-335/test_read_text_remote_args0')

    def test_read_text_remote_args(ray_start_cluster, tmp_path):
        cluster = ray_start_cluster
        cluster.add_node(
            resources={"foo": 100},
            num_cpus=1,
            _system_config={"max_direct_call_object_size": 0},
        )
        cluster.add_node(resources={"bar": 100}, num_cpus=1)
    
        ray.init(cluster.address)
    
        @ray.remote
        def get_node_id():
            return ray.get_runtime_context().node_id.hex()
    
        bar_node_id = ray.get(get_node_id.options(resources={"bar": 1}).remote())
    
        path = os.path.join(tmp_path, "test_text")
        os.mkdir(path)
        with open(os.path.join(path, "file1.txt"), "w") as f:
            f.write("hello\n")
            f.write("world")
        with open(os.path.join(path, "file2.txt"), "w") as f:
            f.write("goodbye")
    
>       ds = ray.data.read_text(
            path, parallelism=2, ray_remote_args={"resources": {"bar": 1}}
        )

python/ray/data/tests/test_dataset_text.py:138: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
python/ray/data/read_api.py:690: in read_text
    return read_binary_files(
python/ray/data/dataset.py:711: in flat_map
    return Dataset(plan, self._epoch, self._lazy)
python/ray/data/dataset.py:217: in __init__
    self._plan.execute(allow_clear_input_blocks=False)
python/ray/data/_internal/plan.py:308: in execute
    blocks, stage_info = stage(
python/ray/data/_internal/plan.py:662: in __call__
    blocks = compute._apply(
python/ray/data/_internal/compute.py:134: in _apply
    metadata = ray.get(refs.pop(-1))
python/ray/_private/client_mode_hook.py:105: in wrapper
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

object_refs = [ObjectRef(7109b8141612f944ffffffffffffffffffffffff0100000003000000)]

    @PublicAPI
    @client_mode_hook(auto_init=True)
    def get(
        object_refs: Union[ray.ObjectRef, Sequence[ray.ObjectRef]],
        *,
        timeout: Optional[float] = None,
    ) -> Union[Any, List[Any]]:
        """Get a remote object or a list of remote objects from the object store.
    
        This method blocks until the object corresponding to the object ref is
        available in the local object store. If this object is not in the local
        object store, it will be shipped from an object store that has it (once the
        object has been created). If object_refs is a list, then the objects
        corresponding to each object in the list will be returned.
    
        Ordering for an input list of object refs is preserved for each object
        returned. That is, if an object ref to A precedes an object ref to B in the
        input list, then A will precede B in the returned list.
    
        This method will issue a warning if it's running inside async context,
        you can use ``await object_ref`` instead of ``ray.get(object_ref)``. For
        a list of object refs, you can use ``await asyncio.gather(*object_refs)``.
    
        Args:
            object_refs: Object ref of the object to get or a list of object refs
                to get.
            timeout (Optional[float]): The maximum amount of time in seconds to
                wait before returning.
    
        Returns:
            A Python object or a list of Python objects.
    
        Raises:
            GetTimeoutError: A GetTimeoutError is raised if a timeout is set and
                the get takes longer than timeout to return.
            Exception: An exception is raised if the task that created the object
                or that created one of the objects raised an exception.
        """
        worker = global_worker
        worker.check_connected()
    
        if hasattr(worker, "core_worker") and worker.core_worker.current_actor_is_asyncio():
            global blocking_get_inside_async_warned
            if not blocking_get_inside_async_warned:
                logger.warning(
                    "Using blocking ray.get inside async actor. "
                    "This blocks the event loop. Please use `await` "
                    "on object ref with asyncio.gather if you want to "
                    "yield execution to the event loop instead."
                )
                blocking_get_inside_async_warned = True
    
        with profiling.profile("ray.get"):
            is_individual_id = isinstance(object_refs, ray.ObjectRef)
            if is_individual_id:
                object_refs = [object_refs]
    
            if not isinstance(object_refs, list):
                raise ValueError(
                    "'object_refs' must either be an ObjectRef or a list of ObjectRefs."
                )
    
            # TODO(ujvl): Consider how to allow user to retrieve the ready objects.
            values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
            for i, value in enumerate(values):
                if isinstance(value, RayError):
                    if isinstance(value, ray.exceptions.ObjectLostError):
                        worker.core_worker.dump_object_store_memory_usage()
                    if isinstance(value, RayTaskError):
                        raise value.as_instanceof_cause()
                    else:
>                       raise value
E                       ray.exceptions.ObjectFetchTimedOutError: Failed to retrieve object 7109b8141612f944ffffffffffffffffffffffff0100000003000000. To see information about where this ObjectRef was created in Python, set the environment variable RAY_record_ref_creation_sites=1 during `ray start` and `ray.init()`.
E                       
E                       Fetch for object 7109b8141612f944ffffffffffffffffffffffff0100000003000000 timed out because no locations were found for the object. This may indicate a system-level bug.

python/ray/_private/worker.py:2281: ObjectFetchTimedOutError
```

### Versions / Dependencies

Master

### Reproduction script

As above.

### Issue Severity

High: It blocks me from completing my task.

--- a/python/ray/tests/test_generators.py
+++ b/python/ray/tests/test_generators.py
@@ -194,15 +194,9 @@ def test_dynamic_generator_distributed(ray_start_cluster):
             yield np.ones(1_000_000, dtype=np.int8) * i
             time.sleep(0.1)
 
-    @ray.remote
-    def read(gen):
-        for i, ref in enumerate(gen):
-            if ray.get(ref)[0] != i:
-                return False
-        return True
-
     gen = ray.get(dynamic_generator.remote(3))
     for i, ref in enumerate(gen):
+        # Check that we can fetch the values from a different node.
         assert ray.get(ref)[0] == i
 
 

--- a/python/ray/tests/test_generators.py
+++ b/python/ray/tests/test_generators.py
@@ -1,6 +1,7 @@
 import pytest
 import numpy as np
 import sys
+import time
 
 import ray
 
@@ -177,6 +178,32 @@ def test_dynamic_generator(ray_start_regular, store_in_plasma):
 
     with pytest.raises(ray.exceptions.RayTaskError):
         ray.get(static.remote(3))
+
+
+def test_dynamic_generator_distributed(ray_start_cluster):
+    cluster = ray_start_cluster
+    # Head node with no resources.
+    cluster.add_node(num_cpus=0)
+    ray.init(address=cluster.address)
+    cluster.add_node(num_cpus=1)
+    cluster.wait_for_nodes()
+
+    @ray.remote(num_returns="dynamic")
+    def dynamic_generator(num_returns):
+        for i in range(num_returns):
+            yield np.ones(1_000_000, dtype=np.int8) * i
+            time.sleep(0.1)
+
+    @ray.remote
+    def read(gen):
+        for i, ref in enumerate(gen):
+            if ray.get(ref)[0] != i:
+                return False
+        return True
+
+    gen = ray.get(dynamic_generator.remote(3))
+    for i, ref in enumerate(gen):
+        assert ray.get(ref)[0] == i
 
 
 def test_dynamic_generator_reconstruction(ray_start_cluster):

--- a/src/ray/core_worker/core_worker.cc
+++ b/src/ray/core_worker/core_worker.cc
@@ -3001,8 +3001,8 @@ void CoreWorker::AddObjectLocationOwner(const ObjectID &object_id,
   }
 
   // For generator tasks where we haven't yet received the task reply, the
-  // internal ObjectRefs may not be added yet, don't find out about these until
-  // the task finishes.
+  // internal ObjectRefs may not be added yet, so we don't find out about these
+  // until the task finishes.
   const auto &maybe_generator_id = task_manager_->TaskGeneratorId(object_id.TaskId());
   if (!maybe_generator_id.IsNil()) {
     // The task is a generator and may not have finished yet. Add the internal

--- a/src/ray/core_worker/core_worker.cc
+++ b/src/ray/core_worker/core_worker.cc
@@ -2976,12 +2976,6 @@ void CoreWorker::AddSpilledObjectLocationOwner(
     // object is spilled before the reply from the task that created the
     // object. Add the dynamically created object to our ref counter so that we
     // know that it exists.
-    // NOTE(swang): We don't need to do this for in-plasma object locations because:
-    // 1) We will add the primary copy as a location when processing the task
-    // reply.
-    // 2) It is not possible to copy the object to a second location until
-    // after the owner has added the object to the ref count table (since no
-    // raylet can get the current location of the object until this happens).
     RAY_CHECK(!generator_id->IsNil());
     reference_counter_->AddDynamicReturn(object_id, *generator_id);
   }

--- a/src/ray/core_worker/core_worker.cc
+++ b/src/ray/core_worker/core_worker.cc
@@ -3005,6 +3005,17 @@ void CoreWorker::AddObjectLocationOwner(const ObjectID &object_id,
   if (!reference_exists) {
     RAY_LOG(DEBUG) << "Object " + object_id.Hex() + " not found";
   }
+
+  // For generator tasks where we haven't yet received the task reply, the
+  // internal ObjectRefs may not be added yet, don't find out about these until
+  // the task finishes.
+  const auto &maybe_generator_id = task_manager_->TaskGeneratorId(object_id.TaskId());
+  if (!maybe_generator_id.IsNil()) {
+    // The task is a generator and may not have finished yet. Add the internal
+    // ObjectID so that we can update its location.
+    reference_counter_->AddDynamicReturn(object_id, maybe_generator_id);
+    RAY_UNUSED(reference_counter_->AddObjectLocation(object_id, node_id));
+  }
 }
 
 void CoreWorker::RemoveObjectLocationOwner(const ObjectID &object_id,

--- a/src/ray/core_worker/task_manager.cc
+++ b/src/ray/core_worker/task_manager.cc
@@ -809,5 +809,17 @@ void TaskManager::FillTaskInfo(rpc::GetCoreWorkerStatsReply *reply,
   reply->set_tasks_total(total);
 }
 
+ObjectID TaskManager::TaskGeneratorId(const TaskID &task_id) const {
+  absl::MutexLock lock(&mu_);
+  auto it = submissible_tasks_.find(task_id);
+  if (it == submissible_tasks_.end()) {
+    return ObjectID::Nil();
+  }
+  if (!it->second.spec.ReturnsDynamic()) {
+    return ObjectID::Nil();
+  }
+  return it->second.spec.ReturnId(0);
+}
+
 }  // namespace core
 }  // namespace ray

--- a/src/ray/core_worker/task_manager.h
+++ b/src/ray/core_worker/task_manager.h
@@ -284,13 +284,9 @@ class TaskManager : public TaskFinisherInterface, public TaskResubmissionInterfa
   /// Fill every task information of the current worker to GetCoreWorkerStatsReply.
   void FillTaskInfo(rpc::GetCoreWorkerStatsReply *reply, const int64_t limit) const;
 
-  /// Update nested ref count info and store the in-memory value for a task's
-  /// return object. Returns true if the task's return object was returned
-  /// directly by value.
-  bool HandleTaskReturn(const ObjectID &object_id,
-                        const rpc::ReturnObject &return_object,
-                        const NodeID &worker_raylet_id,
-                        bool store_in_plasma);
+  /// Returns the generator ID that contains the dynamically allocated
+  /// ObjectRefs, if the task is dynamic. Else, returns Nil.
+  ObjectID TaskGeneratorId(const TaskID &task_id) const;
 
  private:
   struct TaskEntry {
@@ -366,6 +362,16 @@ class TaskManager : public TaskFinisherInterface, public TaskResubmissionInterfa
     // The task's current execution status.
     rpc::TaskStatus status = rpc::TaskStatus::PENDING_ARGS_AVAIL;
   };
+
+
+  /// Update nested ref count info and store the in-memory value for a task's
+  /// return object. Returns true if the task's return object was returned
+  /// directly by value.
+  bool HandleTaskReturn(const ObjectID &object_id,
+                        const rpc::ReturnObject &return_object,
+                        const NodeID &worker_raylet_id,
+                        bool store_in_plasma)
+      LOCKS_EXCLUDED(mu_);
 
   /// Remove a lineage reference to this object ID. This should be called
   /// whenever a task that depended on this object ID can no longer be retried.

--- a/src/ray/core_worker/task_manager.h
+++ b/src/ray/core_worker/task_manager.h
@@ -363,15 +363,13 @@ class TaskManager : public TaskFinisherInterface, public TaskResubmissionInterfa
     rpc::TaskStatus status = rpc::TaskStatus::PENDING_ARGS_AVAIL;
   };
 
-
   /// Update nested ref count info and store the in-memory value for a task's
   /// return object. Returns true if the task's return object was returned
   /// directly by value.
   bool HandleTaskReturn(const ObjectID &object_id,
                         const rpc::ReturnObject &return_object,
                         const NodeID &worker_raylet_id,
-                        bool store_in_plasma)
-      LOCKS_EXCLUDED(mu_);
+                        bool store_in_plasma) LOCKS_EXCLUDED(mu_);
 
   /// Remove a lineage reference to this object ID. This should be called
   /// whenever a task that depended on this object ID can no longer be retried.

